reader.go

```diff
@@ -122,76 +122,56 @@ func (r *Reader) Config() ReaderConfig {
 // messages from it.
 func (r *Reader) Close() error {
 	r.mutex.Lock()
-	if !r.closed {
-		close(r.done)
-	}
+	closed := r.closed
 	r.closed = true
-	r.dirty = false
-	r.cancel()
 	r.mutex.Unlock()
+
+	r.cancel()
+	r.join.Wait()
```

Comment:
we didn't wait before? be nice to have a test showing that we wait for internal goroutines to exit first.